misc-utils/lsfd.c

```diff
@@ -473,85 +386,74 @@ static void collect_fd_files(struct path_cxt *pc, struct proc *proc)
 	}
 }
 
-static struct map* new_map(unsigned long mem_addr_start, unsigned long long file_offset,
-			    int r, int w, int x, int s)
+static void parse_maps_line(char *buf, struct proc *proc)
 {
-	struct map *map = xcalloc(1, sizeof(*map));
-
-	INIT_LIST_HEAD(&map->maps);
+	uint64_t start, end, offset, ino;
+	unsigned long major, minor;
+	enum association assoc = ASSOC_MEM;
+	struct stat sb;
+	struct file *f;
+	char *path, modestr[5];
 
-	map->mem_addr_start = mem_addr_start;
-	map->file_offset = file_offset;
+	/* ignore non-path entries */
+	path = strchr(buf, '/');
+	if (!path)
+		return;
+	rtrim_whitespace((unsigned char *) path);
 
-	map->read   = r;
-	map->write  = w;
-	map->exec   = x;
-	map->shared = s;
+	/* first try the path */
+	if (stat(path, &sb) < 0)
```

Comment:
Though I agreed with you that we don't need `map_file`, this line tells us the limitation of `maps`.
This line tries to get `sb` for `path` in the current mnt namespace.
If the fs tree is rearranged with bind mount after the target process opening a file, lsfd can be cheated.
`map_files` is reliable.

And this may be the reason why map_file is introduced to Linux kernel.

Reply:
Good point. 

Frankly, I'm not sure how does it work with over-mount. This is something we need to test, I think we can detect the problem as the `maps` file contains devno and inode, so we can compare these numbers with the result from stat(), and in case of any problem we can try /proc/#/map_files/xxxx-xxx symlink as a fallback solution. I'll add TODO comment to the code.

The namespaces are another story, I guess readlink() for /proc/#/map_files/xxxx-xxxx is also interpreted in the current namespace. The ideal solution for all files would be to detect that process uses an unshared mount namespace and use setns() to enter the namespace before we start to collect files. 

Definitely something we need to care about, but later ;-)

Reply:
> Frankly, I'm not sure how does it work with over-mount. This is something we need to test, I think we can detect the problem as the maps file contains devno and inode, so we can compare these numbers with the result from stat(), and in case of any problem we can try /proc/#/map_files/xxxx-xxx symlink as a fallback solution. I'll add TODO comment to the code.

When testing, we should add btrfs as a target platform. It seems that a devno on btrfs is not so consistent.
See https://github.com/lsof-org/lsof/issues/152

> The namespaces are another story, I guess readlink() for /proc/#/map_files/xxxx-xxxx is also interpreted in the current namespace. 

As you wrote, if you pass the path getting from readlink() to stat(), it doesn't make sense.
However, I expect doing fstatat /proc/#/map_files/xxxx-xxx doesn't do any filename-resolution.
I will test it.

> Definitely something we need to care about, but later ;-)

Yes.